Add `download-xsd` CLI command

leebradley opened the pull request:
Add an XSD download CLI command, along with several utility methods for
downloading, extracting, and validating the files within the schema
archive.

Signed-off-by: Lee Bradley <bradley@bitwise.io>

dnewh reviewed (dismissed):
tested

CLEckhardt commented on cli/src/actions/xsd_downloader/validator.rs:
```diff
@@ -0,0 +1,77 @@
+pub fn validate_hash(file: &Path, hash: &[u8; 32]) -> Result<(), CliError> {
+    info!("validating hash of {file}", file = file.to_string_lossy());
+    let mut hasher = Sha256::new();
+    let data = std::fs::read(file).unwrap();
```
--no-download panics when there isn't an artifact

```
root@gridd-alpha:/usr/share/grid/xsd# grid download-xsd --no-downl│
oad                                                               │
validating hash of /var/lib/grid/artifact/gs1.zip                 │
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` v│
alue: Os { code: 2, kind: NotFound, message: "No such file or dire│
ctory" }', cli/src/actions/xsd_downloader/validator.rs:15:36      │
note: run with `RUST_BACKTRACE=1` environment variable to display │
a backtrace      
```

leebradley replied:
Good catch! Fixed

CLEckhardt commented on cli/src/main.rs:
```diff
@@ -1775,6 +1777,50 @@ fn run() -> Result<(), CliError> {
+                        .help("Do not perform a checksum on the cached file"),
+                )
+                .arg(
+                    Arg::with_name("copy_from")
```
It looks like this should take a value.

leebradley replied:
Added `takes_value(true)`

CLEckhardt commented on cli/src/main.rs:
```diff
@@ -1775,6 +1777,50 @@ fn run() -> Result<(), CliError> {
+                        .help("Replenish the cache from a directory resource and use that"),
+                )
+                .arg(
+                    Arg::with_name("artifact_dir")
```
It looks like this should also take a value.

leebradley replied:
Added `takes_value(true)`

shannynalayna commented on cli/man/grid-download-xsd.1.md:
```diff
@@ -0,0 +1,77 @@
+% GRID-KEYGEN(1) Cargill, Incorporated | Grid
+
+<!--
+  Copyright 2021 Cargill Incorporated
```
This man page should be copyright for `2022`

leebradley replied:
Updated!

shannynalayna commented on cli/src/actions/xsd_downloader/downloader.rs:
```diff
@@ -0,0 +1,419 @@
+use std::fs::{self, File};
```
This file is missing a copyright

leebradley replied:
Added!

shannynalayna commented on cli/src/actions/xsd_downloader/extractor.rs:
```diff
@@ -0,0 +1,264 @@
+use std::fs::{self, File};
```
Need a copyright

leebradley replied:
Added!

shannynalayna commented on cli/src/actions/xsd_downloader/mod.rs:
```diff
@@ -0,0 +1,603 @@
+mod downloader;
```
Copyright :)

shannynalayna replied:
Copyrights need to be added to all the new files, noticed a few more but won't comment everywhere

leebradley replied:
Added!

shannynalayna commented on cli/src/main.rs:
```diff
@@ -1775,6 +1774,54 @@ fn run() -> Result<(), CliError> {
+                    Arg::with_name("copy_from")
+                        .takes_value(true)
+                        .conflicts_with("force_download")
+                        .conflicts_with("no_download")
```
From the man page, it looks like this option should be usable with `no_download`. The man page should be updated to show these args are not usable together, unless this line is a mistake?

leebradley replied:
So actually I'd somehow missed the req that no_download should be compatible with copy_from, so you found a legit bug here. Fixed and added a unit test to validate the functionality.

shannynalayna reviewed (dismissed):
Tested

CLEckhardt reviewed (commented):
Tested and functionality is good. A few notes:
- `--copy-from` without `--no-download` effectively ignores the arg if the filename is incorrect
- `--force` still gives a message “validating hash of ...”
- man page and CLI help should probably say that the file for `--copy-from` needs to be named “gs1.zip”
Still working on testing po actions, running into some unrelated issues, but I'll confirm it all works when that's done

leebradley commented:
> * `--copy-from` without `--no-download` effectively ignores the arg if the filename is incorrect

Yeah, I was initially going to be very strict and require all files exist, but there's also the use case @shannynalayna pointed out of only having some of the files available (and downloading the rest). I'm not sure how to best approach this.

> * `--force` still gives a message “validating hash of ...”

By design `--force` still validates the hash of downloaded files, and only ignores the hashes of cached files.

> * man page and CLI help should probably say that the file for `--copy-from` needs to be named “gs1.zip”

Hmm if this is going to be added to docs that means we'll technically be supporting manually tinkering with the internal cache, which tbh I didn't anticipate. I think I'm going to rename this file so it's a little more future-proof.

CLEckhardt commented:
Cool, makes sense.

leebradley commented:
So in response to the feedback form @CLEckhardt I've made three changes:

1) After consulting a bit, we've decided to support users working with these cache files. Thus I've renamed "gs1.zip" in the cache to "GS1_XML_3-4-1_Publication.zip" so it matches the default filename and will not need renamed in the future if a different version comes out.

2) Added a reference to GS1_XML_3-4-1_Publication.zip in the docs so users are aware what needs to be in the copy directory.

3) Changed a `debug!` message to `info!` so users can immediately see if there are missing files in the cache directory. Example below:

```
root@gridd-alpha:/# grid download-xsd --copy-from foo              
file "GS1_XML_3-4-1_Publication.zip" doesn't exist in copy path
downloading file from https://www.gs1.org/docs/EDI/xml/3.4.1/GS1_XML_3-4-1_Publication.zip
download finished
validating hash of /var/cache/grid/xsd_artifact_cache/GS1_XML_3-4-1_Publication.zip.download
extracting to schema directory
```

shannynalayna commented on cli/man/grid-download-xsd.1.md:
```diff
@@ -0,0 +1,79 @@
+This command downloads GS1 XSD files used by various Grid features. The
+downloaded artifacts are first copied into a cache directory. They are then
+expanded into Grid's state directory. If the desired artifacts are in the
+cahce directory, Grid will not attempt to re-download them, and instead
```
sp/cahce/cache

leebradley replied:
Fixed!

shannynalayna reviewed (approved):
Tested; temp file is being removed if the overall download fails
